### DISCUSSION TITLE
Feature request: give consult-buffer the ability to only consider visibility in the current frame

### DISCUSSION THREAD

**Discussion by WJCFerguson** [minad/consult#1082] [Ideas]
I drop emacs frames in various Window Manager workspaces and would like them to act more independently with respect to switching buffers.  As it is, the sort will bury buffers that are visible in a frame in another WM workspace (even if they're not on screen).

I would submit a PR but I'm pretty unsure about what is the *right* way to do it.  I hacked it for my personal use by copying `defun consult--buffer-sort-frame-visible` from the `...-sort-visible` defun so it calls `get-buffer-window` with `(selected-frame)` as its third arg, and then tweaking `consult--source-buffer` to make `consult--buffer-query` use that sort function.

``` elisp
  (defun consult--buffer-sort-frame-visible (buffers)
    "Sort BUFFERS by visibility in current frame.

Need this function so consult--source-buffer can be configured to
call consult--buffer-query with sort of 'frame-visible."
    (let ((hidden)
          (current (car (memq (current-buffer) buffers))))
      (consult--keep! buffers
        (unless (eq it current)
          (if (get-buffer-window it (selected-frame))
              it
            (push it hidden)
            nil)))
      (nconc (nreverse hidden) buffers (and current (list current)))))

  (setq consult--source-buffer
        `(:name     "Buffer"
                    :narrow   ?b
                    :category buffer
                    :face     consult-buffer
                    :history  buffer-name-history
                    :state    ,#'consult--buffer-state
                    :default  t
                    :items
                    ,(lambda () (consult--buffer-query :sort 'frame-visible
                                                       :as #'consult--buffer-pair))))
```

**Comment by minad** [comment]
Hi! I have a similar problem myself with EXWM where I want to use different tabs independently. See also #1009. However I am not yet sure about the optimal solution, which is flexible enough for the huge variety of possible use cases, where buffers should be organized by frames/tabs/projects/whatever.

**Comment by WJCFerguson** [comment]
Good point on more use cases.  Perhaps an ability to swap in a function alternative to the call to `get-buffer-window` would be a modest/tidy change to consult, but flexible for many uses?

**Reply by minad** [reply]
I am not sure. The result should also be sufficiently efficient. Currently I use the following code for tab/frame isolation:

~~~ elisp
  (setf (plist-get consult--source-buffer :items)
        (lambda ()
          (consult--buffer-query
           :sort 'visibility
           :as #'consult--buffer-pair
           :buffer-list (frame-parameter nil 'buffer-list)))) ;; NOTE THIS
~~~
